### DISCUSSION TITLE
Peek anti-pattern trap?

### DISCUSSION THREAD

**Discussion by TheNeikos** [winnow-rs/winnow#623] [Ideas]
I feel like I accidentally walked into a trap wrt. `Parser::parse_peek`.

What I wanted to accomplish:

- Parse something, if it succeeds continue, if it doesn't, restore the initial input.

`parse_peek` seemed like the correct choice, it returns the rest of the input if it succeeds! This feels like the nom way, so I can continue using that, shadowing the original 'input'.

This happened:

- The caller of my parser never saw progress beyond the `parse_peek`

After some thinking, this made sense, as the 'rest of input' returned by parse_peek is a _copy_ of the original input! Consuming that, does not advance the original of course.

What was the actual solution:

- Using `opt`. I thought through it methodically, and was thinking that this must already exist the in the API. I felt like `opt` was the right choice, and it does do what I wanted it to do!

----

I am not sure if there is anything that _should_ be changed. But I feel like this might happen to others? Maybe I've got a somewhat rare path to using winnow, as I was heavily using nom before. But I feel that by for example following papers or parsers in other languages, the concept of `peek` might be confusing in this situation.

Since lifetimes in Rust are not visible outside of signatures, I feel that this is also a somewhat confusing and hard to debug problem if I hadn't just changed that single piece of code.

Maybe some extra docs on `parse_peek` could alleviate the issue? I also feel that `peek` could get a similar treatement.

My suggestion would be something akin to:

> The remaining input as returned by `parse_peek` is a copy from the input given as an argument. This means that consuming the remaining input **will not** advance the original input! If you would like to consume it, but only on success, you can use `opt`.`

And something similar for `peek`?

Either way, this is more of a "I've used it, felt like there's a rough edge so I'll say something" kind of report. So no problem if things stay as they are!

**Comment by epage** [comment]
Huh, the docs aren't the greatest.  I've tried to improve them in #624 and would appreciate feedback!

As for your journey, there are parts that confuse me.  I'm not calling these out to attack you or to diminish the journey you went through but in better understanding the path someone goes through, we can find new paths through the API / documentation to help people.

> > Parse something, if it succeeds continue, if it doesn't, restore the initial input.
>
> parse_peek seemed like the correct choice, it returns the rest of the input if it succeeds! This feels like the nom way, so I can continue using that, shadowing the original 'input'.
>
> ...
>
> Using opt. I thought through it methodically, and was thinking that this must already exist the in the API. I felt like opt was the right choice, and it does do what I wanted it to do!

Before this, if someone had asked me what the `nom` way of accomplishing that, I would have answered with [`nom::combinator::opt`](https://docs.rs/nom/latest/nom/combinator/fn.opt.html).

My theories as to why that wasn't the answer in your experience are
- nom's immutable input handling makes it trivial to get the behavior of `opt`, maybe even cleaner than `opt` for imperative code
- the structure and extent of nom's API makes it hard to know what all features exist.

> But I feel that by for example following papers or parsers in other languages, the concept of peek might be confusing in this situation.

Have any examples?  We adopted the names `opt` and `peek` from nom and they seemed decent enough to me but we are all for working to find the best names for things (e.g. #95) as it can make a big difference in [approachability and correctness](https://epage.github.io/blog/2023/02/winnow-toml-edit-combine-nom/#usage).

> > Parse something, if it succeeds continue, if it doesn't, restore the initial input.
>
> parse_peek seemed like the correct choice, it returns the rest of the input if it succeeds! 
>
> ...
>
> Since lifetimes in Rust are not visible outside of signatures, I feel that this is also a somewhat confusing and hard to debug problem if I hadn't just changed that single piece of code.

For myself, it seems like a problem would be "obvious" but it obviously wasn't for you, so I must be missing something.

If you are writing a parser that calls `parse_peek`, you'd get
```rust
fn expression(input: &mut &str) -> PResult<Expr> {
    // ...
    let (input, output) = identifier.parse_peek(*input)?;
    Ok(output)
}
```

Between `*input`, or even `input.clone()` for a more complex type, conveys that a copy is being made so anything we do afterwards doesn't affect the original `input`.

We would also get a warning about `input` being unused.  For me, seeing the input returned, I would assume that its for a reason and that that is what was changed.

If we changed this to not being in a parser
```rust
let mut input = //

let output = expression.parse_next(&mut input);
// vs
let (input, output) = expression.parse_peek(input);
```

Rust makes it explicit when a function is able to modify `input` or not.

Any insight into where things didn't work out for you?

**Reply by TheNeikos** [reply]
> Any insight into where things didn't work out for you?

My code looked somewhat like this:

```rust
fn parse_expr(input: &mut Str) -> PResult<Expr> {
  // .. do stuff with input
  
  loop {
    let (mut input, val) = match parse_expr.parse_peek(input) {
      // break on backtrack, return on Cut, pass through Ok
    };
    
    // several more lines where I do stuff with `input`
  }
}
```

Calling `parse_expr.parse()` would return an error, as it would not hit EOF, despite having the appearance that I'm fine.

Of course the `mut` here is a bit suspicious, but since parsers take `&mut I` this wasn't suuper out of the ordinary.

> Have any examples?

In this case I was following this [pratt parsing blog post](https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html). 

In my mind, 'peeking' is parsing without advancing the input. So I was thinking that the function that had those names where what I needed. However, with either `peep` or `parse_peek` there was no clear way for me to then go "Ok, so what I peeked is fine. Please advance the input now."

Using a mix of `opt(parser.verify)` I was able to get what I wanted in the end. But I first had to let go of the idea of 'peeking' and rather identify it with it being a check/revert kind of parser.

Maybe the concept could also be called something like 'try'? Which I also tried searching for, but there's nothing like that. (Maybe one could use the docs annotation for alternative names here? They help when searching.)

But in general, I think that I was basically led astray by `parse_peek`... existing? #624 would have made it clear that I was on the wrong path, and maybe gotten me to the better choice faster.

I think that extending that PR with links to some alternative parsers would make it even better!

**Reply by epage** [reply]
> My code looked somewhat like this:

Huh, I thought `*input` would have been needed.

**Reply by epage** [reply]
> I think that extending that PR with links to some alternative parsers would make it even better!

I had considered that but was unsure how to frame the relationship of `peek` and `opt`.  However, the way you did of "peeking and then committing on success" works.
